connectors/spark/src/main/scala/io/unitycatalog/spark/AwsVendedTokenProvider.java

```diff
@@ -0,0 +1,125 @@
+package io.unitycatalog.spark;
+
+import io.unitycatalog.client.ApiClient;
+import io.unitycatalog.client.ApiException;
+import io.unitycatalog.client.api.TemporaryCredentialsApi;
+import io.unitycatalog.client.model.*;
+import org.apache.hadoop.conf.Configuration;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+
+import java.net.URI;
+
+public class AwsVendedTokenProvider implements AwsCredentialsProvider {
+    // The time remaining until expiration, we will try to renew the credential before the expiration time.
+    private static final long DEFAULT_RENEWAL_LEAD_TIME = 30 * 1000;
+
+    private final Configuration conf;
+    private final URI uri;
+    private final String token;
+
+    private volatile long renewalLeadTime = DEFAULT_RENEWAL_LEAD_TIME;
+
+    private volatile AwsS3Credentials awsS3Credentials;
+    private volatile ApiClient lazyApiClient = null;
+
+    /**
+     * Constructor for the hadoop's CredentialProviderListFactory#buildAWSProviderList to initialize.
+     */
+    public AwsVendedTokenProvider(URI ignored, Configuration conf) {
+        this.conf = conf;
+        this.uri = URI.create(conf.get(Constants.UNITY_CATALOG_URI));
+        this.token = conf.get(Constants.UNITY_CATALOG_TOKEN);
+    }
+
+    private ApiClient apiClient() {
+        if(lazyApiClient == null) {
+            synchronized (this) {
+                if(lazyApiClient == null) {
+                    lazyApiClient = ApiClientFactory.createApiClient(uri, token);
+                }
+            }
+        }
+
+        return lazyApiClient;
+    }
+
+    @Override
+    public AwsCredentials resolveCredentials() {
+        if(awsS3Credentials == null || awsS3Credentials.readyToRenew()) {
+            synchronized (this) {
+                if(awsS3Credentials == null || awsS3Credentials.readyToRenew()) {
```

Comment:
this provider object will be recreated on every file access (since we disabled FS cache) ...isnt it? so we need to cache the credentials outside this objects lifecycle isnt it?